spec.md

```diff
@@ -336,19 +336,30 @@ on the definition of OPTIONAL.
   event format defines the relationship in
   [section 3.1](./json-format.md#31-special-handling-of-the-data-attribute).
 
-  When this attribute is omitted, the `data` attribute simply follows the event
-  format's encoding rules. For the JSON event format, the `data` attribute value
-  can therefore be a JSON object, array, or value.
-
   For the binary mode of some of the CloudEvents transport bindings, where the
   `data` content is immediately mapped into the payload of the transport frame,
   this field is directly mapped to the respective transport or application
   protocol's content-type metadata property. Normative rules for the binary mode
   and the content-type metadata mapping can be found in the respective transport
   mapping specifications.
 
+  Implementations MAY introduce this parameter when encoding `Integer`, `Map`
+  and `String` to a binary mode of a CloudEvents transport binding. For example,
+  when encoding a `Map` in binary mode, the implementation MAY encode the value
+  as JSON if the underlying transport encoding does not support `Map` payloads,
+  in accordance to the normative rules for the `Map` type, and therefore MUST
+  set the `datacontenttype` to `application/json`. When decoding the event, the
+  new type of `data` will therefore be `Binary`, with a `datacontenttype` of
+  `application/json`.
+
 - Constraints:
-  - OPTIONAL
+  - MUST be compatible with the type of the `data` attribute, given the
+    following rules:
+    - `Binary` - REQUIRED, and can be any media type that can be used to decode
+      the value of the `data` attribute.
+    - `String` - OPTIONAL, and MUST be `text/*` if present.
```

Comment:
I've covered this ground also with the proposal above, but just for completeness of the review:
* There are many "application/*" types that are also text
* Your rule for Map and Integer here (MUST NOT) contradicts the rule above (MAY).

Reply:
The problem with allowing non `"text/*"` types to be string is that when encoding to JSON, we have no other heuristic to know when it should be base64 unencoded or not when it's received. Let's say I am the sender, and I send a `String` with a `datacontenttype` of `application/mytexttype`, and I encode it as JSON, so it gets put as is. Then it gets received by a transcoder that transcodes to protobuf, but that transcoder has no knowledge of my mime type, so assumes it's binary. Because it's binary, it base64 decodes the string, and ends up putting junk in the protobuf binary field - the event is now corrupt. The point is, to prevent such corruptions, we need a robust way for JSON decoders to know if the data is string or binary, only allowing `text/*` is robust, but if we remove that rule, then we need a different way.

The rule for `Map` and `Integer` here doesn't contradict the rule above, because as soon as you encode the `Map` to JSON bytes, it becomes `Binary`, so is no longer a `Map`, and so this rule not applicable.

Reply:
My point is that "text/*" is not a reliable indicator for text content. "application/xml" is text. 

There's no ambiguity around text vs binary because we have a rule that if `Binary` is inside a string, you must set `datacontentencoding` to "Base64". Base64 in a string in `data` cannot exist without that attribute being present. That was what @alanconway raised and we fixed that problem with that attribute.

In binary transport mode, the payload is a sequence of bytes or text as indicated by the media type. In doubt, e.g. if you can't tell what the content type means, it's always binary and if you find a "charset" parameter, you know that you ought to run this through a character encoder that yields text. 

The Map/Integer rules are contradictory because the JSON event format retains the ability to inline JSON content as JSON into the `data` attribute. The content only turns into binary if you transcode the outer event into something else. Otherwise, `data` just carries a JSON value and with the declaration in place, the JSON may also contain bool or values that deviate from our rule set. Without the declaration, you are limited to the abstract type set.